src/Graphics/Glyphs/GlyphGeom.cc

```diff
@@ -523,228 +567,632 @@ void GlyphGeom::generateSphere(const Point& center, double radius, int resolutio
+    double temp = eigvals[1];
+    eigvals[1] = eigvals[2];
+    eigvals[2] = temp;
+    Vector tempVec = eigvectors[1];
+    eigvectors[1] = eigvectors[2];
+    eigvectors[2] = tempVec;
+  }
+  if(eigvals[0] < eigvals[1])
+  {
+    double temp = eigvals[0];
+    eigvals[0] = eigvals[1];
+    eigvals[1] = temp;
+    Vector tempVec = eigvectors[0];
+    eigvectors[0] = eigvectors[1];
+    eigvectors[1] = tempVec;
+  }
 }
 
-void GlyphGeom::generateEllipsoid(const Point& center, Tensor& t, Vector &scaled_eigenvals, int resolution, const ColorRGB& color)
+void GlyphGeom::generateBox(const Point& center, Tensor& t, double scale, ColorRGB& node_color, bool normalize)
 {
-    Vector eig_vec1, eig_vec2, eig_vec3;
-    t.get_eigenvectors(eig_vec1, eig_vec2, eig_vec3);
+  double zeroThreshold = 0.000001;
+  double eigval1, eigval2, eigval3;
+  t.get_eigenvalues(eigval1, eigval2, eigval3);
+
+  Vector eigvals(abs(eigval1), abs(eigval2), abs(eigval3));
+  if(normalize)
+    eigvals.normalize();
+  eigvals *= scale;
+
+  std::vector<Vector> eigvectors(3);
+  t.get_eigenvectors(eigvectors[0], eigvectors[1], eigvectors[2]);
+
+  // Checks to see if eigenvalues are close to 0
+  bool eig_x_0 = eigvals.x() <= zeroThreshold;
+  bool eig_y_0 = eigvals.y() <= zeroThreshold;
+  bool eig_z_0 = eigvals.z() <= zeroThreshold;
+
+  // Set to 0 if below threshold
+  eigvals[0] = (!eig_x_0) * eigvals[0];
+  eigvals[1] = (!eig_y_0) * eigvals[1];
+  eigvals[2] = (!eig_z_0) * eigvals[2];
+
+  bool flatTensor = (eig_x_0 + eig_y_0 + eig_z_0) >= 1;
+  if(flatTensor)
+  {
+    reorderTensor(eigvectors, eigvals);
+
+    eig_x_0 = eigvals.x() <= zeroThreshold;
+    eig_y_0 = eigvals.y() <= zeroThreshold;
+    eig_z_0 = eigvals.z() <= zeroThreshold;
+    // Check for zero eigenvectors
+    if(eig_x_0)
+    {
+      eigvectors[0] = Cross(eigvectors[1], eigvectors[2]);
+    }
+    else if(eig_y_0)
+    {
+      eigvectors[1] = Cross(eigvectors[0], eigvectors[2]);
+    }
+    else if(eig_z_0)
+    {
+      eigvectors[2] = Cross(eigvectors[0], eigvectors[1]);
+    }
+  }
+
+  Transform rotate(Point(0.0, 0.0, 0.0), eigvectors[0], eigvectors[1], eigvectors[2]);
+  Transform trans = rotate;
+  trans.pre_translate((Vector) center);
+
+  // Rotate and translate points
+  Vector p1 = Vector(trans * Point(eigvals * Vector(-1.0, 1.0, 1.0)));
+  Vector p2 = Vector(trans * Point(eigvals * Vector(-1.0, 1.0, -1.0)));
+  Vector p3 = Vector(trans * Point(eigvals * Vector(1.0, 1.0, 1.0)));
+  Vector p4 = Vector(trans * Point(eigvals * Vector(1.0, 1.0, -1.0)));
+  Vector p5 = Vector(trans * Point(eigvals * Vector(-1.0, -1.0, 1.0)));
+  Vector p6 = Vector(trans * Point(eigvals * Vector(-1.0, -1.0, -1.0)));
+  Vector p7 = Vector(trans * Point(eigvals * Vector(1.0, -1.0, 1.0)));
+  Vector p8 = Vector(trans * Point(eigvals * Vector(1.0, -1.0, -1.0)));
+
+  // Rotate norms
+  Vector x_vec = rotate * Vector(1, 0, 0);
+  Vector y_vec = rotate * Vector(0, 1, 0);
+  Vector z_vec = rotate * Vector(0, 0, 1);
```

Comment:
This block can be refactored into the `BoxBuilder` class as well.
```
auto xyz = bb.makeNorms(rotate);
//...
generateBoxSide(box[6], box[7], box[2], box[3], xyz[0], node_color);
```

Reply:
I changed it to get the vectors from the transform and avoided the multiplication entirely.